subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultModuleVersionIdentifier.java

```diff
@@ -23,19 +23,26 @@
 
     private final ModuleIdentifier id;
     private final String version;
+    private final int hashCode;
 
     public DefaultModuleVersionIdentifier(String group, String name, String version) {
         assert group != null : "group cannot be null";
         assert name != null : "name cannot be null";
         assert version != null : "version cannot be null";
         this.id = DefaultModuleIdentifier.newId(group, name);
         this.version = version;
+        // pre-compule the hashcode as it's going to be used anyway, and this object
```

Comment:
I don't find such comments helpful. Precomputing hashcodes is common practice for value objects.

An even more common pattern is computing it on-demand on the first access (see String.java)

Reply:
Well that's actually _why_ I added the comment. I could have computed it lazily, but then it would have implied that the object hashcode may not even be required (which is not the case here, it's _always_ used) and potentially slower (because of thread-safety).

Reply:
There is no thread safety problem when computing it lazily, see String.java.

I still think this comment doesn't add any new insight. If I see this code I know that the author must have done it for that reason.

Reply:
Yeah I said "potential", depending on how it's implemented (for hashcode it probably doesn't matter that 2 threads compute the same value concurrently).

Reply:
I also do not think the comment adds much value. However, in this case, having a final field makes the intent clearer and thus I prefer it over lazy computation.

Reply:
Removed or reworked